sql: fix the batching behavior of mutation statements, especially with RETURNING

knz opened the pull request:
Fixes #22304.
Fixes #23156.
Fixes #24928.

To review: 
- @RaduBerinde for the physical property propagation. Check the order_by test, confirm that this is enables optimizations of DELETE in a WITH clause or [ ... ]
- @jordanlewis for overall design of the new `batchedPlanNode` interface.
- @andreimatei @jordanlewis  for correctness of the batching behavior (like we discussed - I think this does everything you want)

knz committed:
sql: refactor the code for DELETE, fix bugs

Prior to this patch, the plan for the writing statements (DELETE,
INSERT,UPDATE,UPSERT) would have an irregular architecture that
violates the dataflow model used for the local execution of other
plans; namely, the presence of a RETURNING clause would cause the
writing to _push_ rows into a "returningHelper" to prepare its
results; and deleteNode.Values() would defer to
returningHelper.Values().

So instead of the expected call / data flow:

    parent -> returning.Next() -> writingNode.Next()
    parent -> returning.Values() -> writingNode.Values()

we would observe:

    parent -> writingNode.Next()
                -> prepare some updates in a KV batch, do some reads
                -> call returning.cookResultRow() based on the
		   read results, to prepare a RETURNING result row
		-> every other call to Next(), commit the current KV
                   batch and make a new batch.
    parent -> writingNode.Values() -> returning.Values()

This irregularity is common to DELETE, INSERT and UPDATE.  It is not
just an accident of history that results in code merely difficult to
comprehend; it constitutes an outright bug.

The problem can be recognized by stating the obvious: RETURNING
defines a projection over some arbitrary logical plan, and **the
results of RETURNING are only correct assuming that the rows provided
by its source plan are correct.**

Now, consider the pseudo-code above: each call to Next() can make rows
available as RETURNING results, *even if the current KV batch has not
committed yet*. So for example an INSERT of a duplicate row on a
UNIQUE index can return the duplicate row via RETURNING before the
duplicate check occurs (it occurs only when the KV batch is ran).

Worse yet, if a LIMIT is applied on the outside (e.g. via a CTE),
the last batch may not be run at all, resulting in the writing work
not being committed.

This problem needs to be fixed in every write statement, ensuring that
the write work and the checks are completed by the time values are
available via Next()/Values().

This fix requires some careful re-organizing of the code for these
write statements.

This patch _starts_ this work as follows:

- lies the foundation of the fix by proposing an improved logical plan
  structure for writing statements, which comprehensively addresses
  the root problem above.
- demonstrates this foundation by adapting the code for DELETE and
  fixing the corresponding DELETE bugs.

The full fix will require, in later patches, to also adapt the code
for UPDATE and INSERT to use the same foundation, then remove the code
that becomes obsolete as a result. I chose to not perform the fix in
the same patch to reduce the scope of this change and make the review
more manageable.

Proposed foundation:

- make the writing statements implement a new interface
  `batchedPlanNode` that complements `planNode` as follows:

  - its `BatchedNext()` method advances a batch, not a single row anymore.
    This is also guaranteed to run the underlying KV batch(es) to completion.
  - the `BatchedCount()` method returns the number of rows processed
    by the last call to `Next()`.
  - the `BatchedValues()` method gives optional access to a row
    container of batched rows; the rows are guaranteed to be
    valid. There may be no row container if the plan need not produce
    values (e.g. there was no RETURNING clause)

- introduce a "serialize" operator that transforms a `batchedPlanNode` into
  a simple `planNode` suitable by consumption by row-oriented other nodes.

- handle RETURNING clauses using a separate planNode object; in
  particular use `renderNode` with the aforementioned "serialize"
  operator, instead of `returningHelper` when there are expressions to
  render.

The code for DELETE is adapted to use this new model. This removes its
use of the `editNode`, which is fundamentally row-oriented and thus
incompatible with batching.

Subsequent patches will extend the INSERT and UPDATE code accordingly.

Perf changes for DELETE in this patch:

```
name                                old time/op    new time/op    delta
SQL/Cockroach/Delete/count=1-16        428µs ±39%     384µs ± 2%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Delete/count=10-16       512µs ± 1%     515µs ± 2%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Delete/count=100-16     1.85ms ± 9%    1.90ms ± 7%   ~     (p=0.690 n=5+5)
SQL/Cockroach/Delete/count=1000-16    15.8ms ±35%    15.4ms ±24%   ~     (p=1.000 n=5+5)

name                                old alloc/op   new alloc/op   delta
SQL/Cockroach/Delete/count=1-16       58.7kB ± 1%    58.6kB ± 1%   ~     (p=1.000 n=5+5)
SQL/Cockroach/Delete/count=10-16      71.8kB ± 2%    71.5kB ± 1%   ~     (p=0.841 n=5+5)
SQL/Cockroach/Delete/count=100-16      218kB ± 3%     219kB ± 3%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Delete/count=1000-16    1.65MB ± 3%    1.67MB ± 3%   ~     (p=0.421 n=5+5)

name                                old allocs/op  new allocs/op  delta
SQL/Cockroach/Delete/count=1-16          523 ± 0%       526 ± 1%   ~     (p=0.286 n=5+5)
SQL/Cockroach/Delete/count=10-16         638 ± 1%       650 ± 5%   ~     (p=0.159 n=4+5)
SQL/Cockroach/Delete/count=100-16      1.82k ±21%     1.85k ±17%   ~     (p=0.841 n=5+5)
SQL/Cockroach/Delete/count=1000-16     12.8k ±15%     14.7k ±25%   ~     (p=0.222 n=5+5)
```

Release note (bug fix): Fixed a bug where `SELECT * FROM [DELETE FROM
... RETURNING ...] LIMIT 1` or `WITH d AS (DELETE FROM ... RETURNING
...) SELECT * FROM d LIMIT 1` would fail to properly delete some rows.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23373)
<!-- Reviewable:end -->

knz commented:
I am willing to extend this PR with the two subsequent commits that fix UPDATE and INSERT accordingly; or create separate PRs, as you recommend.

In any case I'd like to start the review with the DELETE changes in here, which show the general idea I want to apply to the other things afterwards.

RaduBerinde commented:
Phys props stuff LGTM

---

Review status: 0 of 19 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-:-L6rtDyMHioKi2KxjBR8:bvkfffd)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/plan_batch.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wcl9Ohh17zeuLkpic:-L6wcl9Ohh17zeuLkpid:b-4n1atj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L23)):*
> ```Go
> )
> 
> // batchedPlanNode defines the interface for executing a query
> ```

if this extends `planNode`, then please call this out in the comment.

But I don't really get it - why extend `planNode`, if `Values()` is invalid? Also comments like "foo is contractually re-defined here to..." don't really make sense; that's not how interfaces work :). The whole point is that no implementer or derived interface can change any contract (or at least not in ways that can break unsuspecting clients of the interface).

I think we either extend `planNode`, in which case `Values()` needs to work and the implementers need to buffer rows, or we don't. You can extract all the `planNode` methods except `Next()` and `Values()` into a `planNodeBase` that both `planNode` and `batchedPlanNode` extend.

---

*[pkg/sql/plan_batch.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wcX_UtgexVV4wrsaE:-L6wcX_UtgexVV4wrsaF:b-1qdzsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L29)):*
> ```Go
> 
> 	// The planNode.Next() method is contractually (re-)defined here to
> 	// advance a full batch. Its first (boolean) return value is
> ```

I think "advance a full batch" here and "the number of rows processed in the last call to Next()" in the comment on `BatchedCount` don't mean much.

---

*[pkg/sql/plan_batch.go, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6we3nlo8bZgEBPQqRK:-L6we3nlo8bZgEBPQqRL:b5gc2k9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L33)):*
> ```Go
> 	//
> 	// Additionally, there is no open KV batch (in particular, no
> 	// uncommitted in-progress KV write operation), nor unchecked or
> ```

"uncommitted" is the wrong term to use.

Generally, I don't like this comment very much. It tries to be very general, but I think that a) it will not make much sense to a reader and b) it's too general. It's unclear what "inconsistent" means, it's unclear what "observable" means...
I think we should put more narrow guidance for how writes behave in the face of constraints. Like:
```
NOTE: Nodes that perform writes and also return rows (e.g. INSERT... RETURNING...) do not return rows before checking foreign key, uniqueness and other CHECK constraints. The implementations execute the KV requests for writing the respective rows (which implicitly checks for some of the constraints) and otherwise perform all other checks before returning a row.
```

---

*[pkg/sql/plan_batch.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wg0Uc3EXU8pB2U-UP:-L6wg0Uc3EXU8pB2U-UQ:bo3hnrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L51)):*
> ```Go
> var _ batchedPlanNode = &deleteNode{}
> 
> // serializeNode serializes the results of a batchedPlanNode into a
> ```

nit: I think `batchNodeAdapter` would be a better name. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 19 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/delete.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wjD8_48xQr3RmuZfo:-L6wjD8_48xQr3RmuZfp:b1ryd9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/delete.go#L161)):*
> ```Go
> 	rowsNeeded rowsNeeded
> 
> 	// fastPath is computed during startExec.
> ```

please also say what this `fastPath` means

---

*[pkg/sql/delete.go, line 178 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wjSXbdxWYE8Jx7JmR:-L6wjSXbdxWYE8Jx7JmS:bghb9s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/delete.go#L178)):*
> ```Go
> // single DELETE batch. This combines with maxDeleteBatchSize below
> // to limit the amount of work done in each DELETE batch.
> const maxDeleteBatchRowCount = 1000
> ```

why do we have both these limits? Would one (either one) do by itself?

---

*[pkg/sql/returning.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6whHmBTaQ4EAHAYkUD:-L6whHmBTaQ4EAHAYkUE:b-y6t9t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/returning.go#L27)):*
> ```Go
> )
> 
> // resultsNeeded determines whether a statement that has a RETURNING
> ```

nit: this comment is confusing because it reads as if it was already determined that a RETURNING clause is specified. Say something like "a statement that might have a returning clause"...

---

*[pkg/sql/returning.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6whl7dZG5fkK7C4wfm:-L6whl7dZG5fkK7C4wfn:b-c0yhy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/returning.go#L41)):*
> ```Go
> }
> 
> type rowsNeeded bool
> ```

nit: my preference is to only use bool enums for input arguments. As far as I see, this `rowsNeeded` type is only returned or stored, never passed around, so I think it does more harm than good; a bool would be clearer and would lead to shorter `if`s.

---

*[pkg/sql/returning.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wiIFL2IcgkpQXdPK2:-L6wiIFL2IcgkpQXdPK3:b-ugi0uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/returning.go#L56)):*
> ```Go
> ) (planNode, error) {
> 	// serialize the data-modifying plan to ensure that no data is
> 	// observed that hasn't been committed/checked first.
> ```

"committed" is the wrong word

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/delete.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wjD8_48xQr3RmuZfo:-L7-15FJiW_8OpGI3GaV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/delete.go#L161)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please also say what this `fastPath` means
</blockquote></details>

Done.

---

*[pkg/sql/delete.go, line 178 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wjSXbdxWYE8Jx7JmR:-L7-15lDsqinzZM9Yzfz:bfrii4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/delete.go#L178)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why do we have both these limits? Would one (either one) do by itself?
</blockquote></details>

Yeah I guess just the batch size is important. The row count is constrained by the max batch size anyways. Done.

---

*[pkg/sql/plan_batch.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wcl9Ohh17zeuLkpic:-L7-9ekQdr91krQ6E7Qw:b4kig32) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L23)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if this extends `planNode`, then please call this out in the comment.

But I don't really get it - why extend `planNode`, if `Values()` is invalid? Also comments like "foo is contractually re-defined here to..." don't really make sense; that's not how interfaces work :). The whole point is that no implementer or derived interface can change any contract (or at least not in ways that can break unsuspecting clients of the interface).

I think we either extend `planNode`, in which case `Values()` needs to work and the implementers need to buffer rows, or we don't. You can extract all the `planNode` methods except `Next()` and `Values()` into a `planNodeBase` that both `planNode` and `batchedPlanNode` extend.
</blockquote></details>

Yes I fully agree with you on both points. 
I have tried to implement this but it makes the patch much more invasive and I dislike this amount of changes so late in the 2.0 release. Consider:
- if we keep the base interface called "planNode" used in logical planning, and introduce a new "planNodeRun" that provides Next/Values, I'd need to introduce downcasts from planNode to planNodeRun in all the Next() methods for every planNode implementer.
- if we introduce a new interface "planNodeBase" like you suggest, the local execution code stays unchanged but I need to rewrite all the logical planning code to use that instead.

In both cases the change is rather invasive. I will do it on `master` in a follow-up PR but for this current PR instead I am leaving a TODO and a reference to the new tracking issue #23522.

Also I agree that it is confusing that batchedPlanNode do implement Next() but with a different contract. I have decided to simply make the base method unavailable and instead defined a new `BatchedNext()`.

---

*[pkg/sql/plan_batch.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wcX_UtgexVV4wrsaE:-L7-Dy8jh-pcsJ9VBhl9:b-nfdnca) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L29)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think "advance a full batch" here and "the number of rows processed in the last call to Next()" in the comment on `BatchedCount` don't mean much.
</blockquote></details>

Yes, agreed. Rephrased.

---

*[pkg/sql/plan_batch.go, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6we3nlo8bZgEBPQqRK:-L7-DqW7CEK4GgGMlDyu:bhvv7l1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L33)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

"uncommitted" is the wrong term to use.

Generally, I don't like this comment very much. It tries to be very general, but I think that a) it will not make much sense to a reader and b) it's too general. It's unclear what "inconsistent" means, it's unclear what "observable" means...
I think we should put more narrow guidance for how writes behave in the face of constraints. Like:
```
NOTE: Nodes that perform writes and also return rows (e.g. INSERT... RETURNING...) do not return rows before checking foreign key, uniqueness and other CHECK constraints. The implementations execute the KV requests for writing the respective rows (which implicitly checks for some of the constraints) and otherwise perform all other checks before returning a row.
```
</blockquote></details>

Thanks for the suggestion. Adopted. I extended it a bit because the reasoning goes further and also includes row counts, not just result rows.

---

*[pkg/sql/plan_batch.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wg0Uc3EXU8pB2U-UP:-L7-ArwC6HQuBBFHNjDF:b-ged1te) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L51)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I think `batchNodeAdapter` would be a better name. 
</blockquote></details>

1) We have two such adapters (see the code below) - would be nice to distinguish them.

2) I prefer names that (also) describe what the thing does. "xxxAdapter" doesn't say what xxx is adapted to nor how.

---

*[pkg/sql/returning.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6whHmBTaQ4EAHAYkUD:-L7-EN1B5rLrO_74pPft:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/returning.go#L27)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: this comment is confusing because it reads as if it was already determined that a RETURNING clause is specified. Say something like "a statement that might have a returning clause"...
</blockquote></details>

Done

---

*[pkg/sql/returning.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6whl7dZG5fkK7C4wfm:-L7-ERUiax4Ph_la0JIS:bqf4kfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/returning.go#L41)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: my preference is to only use bool enums for input arguments. As far as I see, this `rowsNeeded` type is only returned or stored, never passed around, so I think it does more harm than good; a bool would be clearer and would lead to shorter `if`s.
</blockquote></details>

Thanks for the advice. I wasn't sure. Done.

---

*[pkg/sql/returning.go, line 56 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L6wiIFL2IcgkpQXdPK2:-L7-FG449vqbS1kmCMrn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/returning.go#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

"committed" is the wrong word
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: refactor the code for UPDATE; fix bugs

This patch applies the new pattern introduced by DELETE in the
previous patch and introduces proper batching behavior for UPDATE.

Perf difference:
```
name                                old time/op    new time/op    delta
SQL/Cockroach/Update/count=1-16        404µs ± 1%     406µs ± 1%   ~     (p=0.556 n=5+4)
SQL/Cockroach/Update/count=10-16       668µs ± 2%     679µs ± 3%   ~     (p=0.310 n=5+5)
SQL/Cockroach/Update/count=100-16     3.14ms ± 3%    3.22ms ± 4%   ~     (p=0.413 n=5+4)
SQL/Cockroach/Update/count=1000-16    24.9ms ±16%    24.0ms ± 2%   ~     (p=0.730 n=5+4)

name                                old alloc/op   new alloc/op   delta
SQL/Cockroach/Update/count=1-16       61.0kB ± 0%    61.0kB ± 1%   ~     (p=0.690 n=5+5)
SQL/Cockroach/Update/count=10-16       131kB ± 1%     131kB ± 1%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Update/count=100-16      797kB ± 0%     799kB ± 1%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Update/count=1000-16    7.11MB ± 1%    7.09MB ± 1%   ~     (p=0.310 n=5+5)

name                                old allocs/op  new allocs/op  delta
SQL/Cockroach/Update/count=1-16          603 ± 4%       605 ± 4%   ~     (p=0.841 n=5+5)
SQL/Cockroach/Update/count=10-16         959 ± 6%       916 ± 6%   ~     (p=0.056 n=5+5)
SQL/Cockroach/Update/count=100-16      3.79k ± 7%     3.94k ± 9%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Update/count=1000-16     32.0k ± 9%     29.3k ± 8%   ~     (p=0.056 n=5+5)
```

(insignificant)

Release note (bug fix): Removed a limitation where `UPDATE` would fail
if the number of modified rows was too large.

Release note (bug fix): Fixed a bug where `SELECT * FROM [UPDATE
... RETURNING ...] LIMIT 1` or `WITH d AS (UPDATE ... RETURNING ...)
SELECT * FROM d LIMIT 1` would fail to properly update some rows.

Release note (sql change): The EXPLAIN output for UPDATE statements
now also report the check expressions, for consistency with INSERT.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Please run the delete benchmark just in case

---

Review status: 0 of 19 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/delete.go, line 38 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L70mr4rCThV_AnDVWSt:-L70mr4rCThV_AnDVWSu:b44zfv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3193f03df098a4c087b6a9fa46f070ee83d73b/pkg/sql/delete.go#L38)):*
> ```Go
> type deleteNode struct {
> 	source  planNode
> 	columns sqlbase.ResultColumns
> ```

please put a comment on this `columns` and say something about RETURNING, since otherwise this field is pretty surprising.

---

*[pkg/sql/plan_batch.go, line 23 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L6wcl9Ohh17zeuLkpic:-L70mMNn1hmx6OVF3as0:bb7efny) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L23)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes I fully agree with you on both points. 
I have tried to implement this but it makes the patch much more invasive and I dislike this amount of changes so late in the 2.0 release. Consider:
- if we keep the base interface called "planNode" used in logical planning, and introduce a new "planNodeRun" that provides Next/Values, I'd need to introduce downcasts from planNode to planNodeRun in all the Next() methods for every planNode implementer.
- if we introduce a new interface "planNodeBase" like you suggest, the local execution code stays unchanged but I need to rewrite all the logical planning code to use that instead.

In both cases the change is rather invasive. I will do it on `master` in a follow-up PR but for this current PR instead I am leaving a TODO and a reference to the new tracking issue #23522.

Also I agree that it is confusing that batchedPlanNode do implement Next() but with a different contract. I have decided to simply make the base method unavailable and instead defined a new `BatchedNext()`.

</blockquote></details>

cool

---

*[pkg/sql/plan_batch.go, line 51 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L6wg0Uc3EXU8pB2U-UP:-L70mb1EDdjrmBWVyPLr:b32wtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a52a05294e230db6b040ce5ad94ca6b4606754f3/pkg/sql/plan_batch.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) We have two such adapters (see the code below) - would be nice to distinguish them.

2) I prefer names that (also) describe what the thing does. "xxxAdapter" doesn't say what xxx is adapted to nor how.
</blockquote></details>

ok

---

*[pkg/sql/tablewriter.go, line 737 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L70oEZZ6Fq0Gf1xHWCT:-L70oEZZ6Fq0Gf1xHWCU:b-w047d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3193f03df098a4c087b6a9fa46f070ee83d73b/pkg/sql/tablewriter.go#L737)):*
> ```Go
> }
> 
> func (td *tableDeleter) rotateBatch(ctx context.Context) error {
> ```

`runBatch()` ?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-:-L70oblZ5BkulaSjaial:byejzom)*
<!-- Sent from Reviewable.io -->

knz commented:
I added the benchmark result in the commit message. It's a wash.

---

Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/delete.go, line 38 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L70mr4rCThV_AnDVWSt:-L70qqyp8tB7iE2UwQV4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3193f03df098a4c087b6a9fa46f070ee83d73b/pkg/sql/delete.go#L38)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please put a comment on this `columns` and say something about RETURNING, since otherwise this field is pretty surprising.
</blockquote></details>

Done.

---

*[pkg/sql/tablewriter.go, line 737 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L70oEZZ6Fq0Gf1xHWCT:-L70qs4eDJiC2rR1w9bB:b-u3adk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3193f03df098a4c087b6a9fa46f070ee83d73b/pkg/sql/tablewriter.go#L737)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`runBatch()` ?
</blockquote></details>

It flushes the previous one and then allocates a new one. Isn't that the definition of rotation? We also "rotate" log files in the same way in the log package.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-:-L70qbpp15oPWIIRuKyw:b-s0axsk)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: refactor the code for INSERT; fix bugs

This patch applies the new pattern introduced by DELETE in the
previous patch and introduces proper batching behavior for INSERT.

Perf difference:
```
name                                             old time/op    new time/op    delta
SQL/Cockroach/Insert/count=1-16                     286µs ± 4%     290µs ± 3%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Insert/count=10-16                    627µs ±22%     628µs ±24%   ~     (p=1.000 n=5+5)
SQL/Cockroach/Insert/count=100-16                  5.44ms ± 9%    5.75ms ±10%   ~     (p=0.310 n=5+5)
SQL/Cockroach/Insert/count=1000-16                 60.7ms ± 3%    59.8ms ± 8%   ~     (p=0.730 n=4+5)
SQL/Cockroach/InsertDistinct/count=1-16            3.44ms ±11%    3.62ms ± 3%   ~     (p=0.151 n=5+5)
SQL/Cockroach/InsertDistinct/count=10-16           1.25ms ±69%    1.06ms ±42%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertDistinct/count=100-16          1.53ms ±15%    1.49ms ± 2%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=1-16            492µs ±11%     499µs ±12%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=5-16            689µs ± 5%     680µs ± 3%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=10-16          1.06ms ± 2%    1.08ms ± 6%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=1-16          1.71ms ±11%    1.65ms ± 4%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=5-16          3.84ms ± 2%    3.83ms ± 3%   ~     (p=1.000 n=4+5)
SQL/Cockroach/InsertFK/count=10/nFks=10-16         5.71ms ±16%    5.85ms ±13%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=1-16         11.4ms ±10%    11.0ms ±11%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=5-16         34.2ms ±12%    33.6ms ±14%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=10-16        51.7ms ± 7%    54.6ms ±25%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=1-16         121ms ± 8%     120ms ± 6%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=5-16         357ms ±10%     406ms ±38%   ~     (p=0.310 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=10-16        603ms ± 9%     610ms ±21%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=1-16       571µs ± 1%     605µs ± 8%   ~     (p=0.730 n=4+5)
SQL/Cockroach/InsertSecondaryIndex/count=10-16     3.63ms ± 8%    3.42ms ±10%   ~     (p=0.413 n=4+5)
SQL/Cockroach/InsertSecondaryIndex/count=100-16    31.5ms ±17%    34.3ms ± 2%   ~     (p=0.800 n=3+2)

name                                             old alloc/op   new alloc/op   delta
SQL/Cockroach/Insert/count=1-16                    39.8kB ± 1%    39.9kB ± 0%   ~     (p=0.548 n=5+5)
SQL/Cockroach/Insert/count=10-16                   88.7kB ± 0%    88.7kB ± 2%   ~     (p=0.690 n=5+5)
SQL/Cockroach/Insert/count=100-16                   567kB ± 2%     574kB ± 3%   ~     (p=0.222 n=5+5)
SQL/Cockroach/Insert/count=1000-16                 5.33MB ± 3%    5.26MB ± 2%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertDistinct/count=1-16             356kB ± 2%     358kB ± 5%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertDistinct/count=10-16            366kB ± 5%     363kB ± 1%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertDistinct/count=100-16           439kB ± 5%     432kB ± 2%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=1-16           57.2kB ± 3%    57.8kB ± 4%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=5-16            126kB ± 2%     125kB ± 4%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=10-16           230kB ± 9%     232kB ±11%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=1-16           327kB ±25%     300kB ±44%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=5-16          1.03MB ±45%    0.97MB ±19%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=10-16         1.92MB ±18%    2.02MB ±31%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=1-16         3.19MB ±37%    2.99MB ±44%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=5-16         14.0MB ±20%    11.8MB ±23%   ~     (p=0.151 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=10-16        24.9MB ±17%    26.9MB ±27%   ~     (p=0.421 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=1-16        55.1MB ±12%    53.3MB ±32%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=5-16         189MB ± 7%     190MB ± 6%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=10-16        330MB ± 5%     321MB ± 7%   ~     (p=0.222 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=1-16       279kB ±16%     290kB ±12%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=10-16     1.69MB ± 6%    1.64MB ± 7%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=100-16    15.9MB ±10%    16.7MB ± 1%   ~     (p=0.800 n=3+2)

name                                             old allocs/op  new allocs/op  delta
SQL/Cockroach/Insert/count=1-16                       311 ± 2%       313 ± 1%   ~     (p=0.143 n=5+5)
SQL/Cockroach/Insert/count=10-16                      478 ± 7%       474 ± 5%   ~     (p=0.730 n=5+5)
SQL/Cockroach/Insert/count=100-16                   2.04k ±10%     2.40k ±29%   ~     (p=0.421 n=5+5)
SQL/Cockroach/Insert/count=1000-16                  22.0k ±45%     18.0k ±22%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertDistinct/count=1-16             2.09k ±23%     2.03k ±17%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertDistinct/count=10-16            2.07k ±23%     2.00k ± 2%   ~     (p=0.421 n=5+5)
SQL/Cockroach/InsertDistinct/count=100-16           2.58k ± 8%     2.54k ± 3%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=1-16              479 ± 6%       496 ± 3%   ~     (p=0.310 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=5-16              935 ± 3%       960 ± 6%   ~     (p=0.222 n=5+5)
SQL/Cockroach/InsertFK/count=1/nFks=10-16           1.68k ±18%     1.70k ±18%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=1-16           2.83k ±39%     2.67k ±73%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=5-16           8.19k ±72%     7.36k ±29%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=10/nFks=10-16          15.2k ±26%     16.2k ±43%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=1-16          28.7k ±47%     25.9k ±57%   ~     (p=0.548 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=5-16           126k ±25%      100k ±33%   ~     (p=0.151 n=5+5)
SQL/Cockroach/InsertFK/count=100/nFks=10-16          215k ±23%      240k ±35%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=1-16          554k ±14%      528k ±38%   ~     (p=1.000 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=5-16         1.75M ± 9%     1.77M ± 7%   ~     (p=0.690 n=5+5)
SQL/Cockroach/InsertFK/count=1000/nFks=10-16        2.85M ± 8%     2.72M ±10%   ~     (p=0.222 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=1-16       2.49k ±19%     2.63k ±15%   ~     (p=0.841 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=10-16      14.4k ± 9%     13.9k ± 9%   ~     (p=0.421 n=5+5)
SQL/Cockroach/InsertSecondaryIndex/count=100-16      121k ±13%      129k ± 1%   ~     (p=0.800 n=3+2)
```

Release note (bug fix): Removed a limitation where `INSERT` would fail
if the number of inserted rows was too large.

Release note (bug fix): Fixed a bug where `SELECT * FROM [INSERT
... RETURNING ...] LIMIT 1` or `WITH d AS (INSERT ... RETURNING ...)
SELECT * FROM d LIMIT 1` would fail to properly update some rows.

nvanbenschoten commented:
This is awesome work! Nice job.

---

Reviewed 6 of 19 files at r1, 3 of 7 files at r2, 1 of 1 files at r3.
Review status: 1 of 25 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/delete.go, line 82 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74XpFTGpWvjnXOTHKU:-L74XpFUIAMtEnOUYJf5:bv9qbw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L82)):*
> ```Go
> 
> 	// Find which table we're working on, check the permissions.
> 	desc, err := ResolveExistingObject(ctx, p, tn, true /*required*/, requireTableDesc)
> ```

nit: these `/*comment*/` comments are diverging from our [style guide](https://github.com/cockroachdb/cockroach/blob/master/docs/style.md#inline-comments-for-arguments-to-function-calls), which expect spaces on either side of the comment. Are you ok with changing these (in a separate PR if you'd like) or should we explore amending the style guide?

---

*[pkg/sql/delete.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74YqcnXh3-BN6ioLYo:-L74YqcnXh3-BN6ioLYp:b-a4978r) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L116)):*
> ```Go
> 		// all).
> 
> 		// TODO(dan): This could be made tighter, just the rows needed for RETURNING
> ```

Is there a corresponding issue for this? Seems like it could be a good starter project.

---

*[pkg/sql/delete.go, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74YTuqBrK9UOKhkqpH:-L74YTuqBrK9UOKhkqpI:b-exv066) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L132)):*
> ```Go
> 	tracing.AnnotateTrace()
> 
> 	// Determine the source for the deletion: the rows that are
> ```

These comments are great! It's now much easier to follow each of the steps in this constructor.

---

*[pkg/sql/delete.go, line 134 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74Z7sfZbnAgakVN-7A:-L74Z7sfZbnAgakVN-7B:bvnv6fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L134)):*
> ```Go
> 	// Determine the source for the deletion: the rows that are
> 	// read, filtered, limited, ordered, etc, prior to the deletion. One
> 	// would think there is only so much one wants to do rows prior to a
> ```

`s/to do rows/to do to rows/`

---

*[pkg/sql/plan_batch.go, line 71 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74cqWXlUIHjqTqPD9B:-L74cqWXlUIHjqTqPD9C:b3ci6eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/plan_batch.go#L71)):*
It took me a few times of reading over this sentence before I understood what it was trying to say. In general, I think interface descriptions are easier to understand if they are declarative about their contract (e.g. `all implementations will ...`) instead of trying to describe implementations (e.g. `all implementations do ...`). Let's reword this to

> Note: Nodes that perform writes (e.g. INSERT) will not return from BatchedNext() before checking foreign key, uniqueness, and other CHECK constraints.

---

*[pkg/sql/plan_batch.go, line 93 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74ey46GTP1Fo2wrqkR:-L74ey46GTP1Fo2wrqkS:b-b898gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/plan_batch.go#L93)):*
> ```Go
> // serializeNode serializes the results of a batchedPlanNode into a
> // plain planNode interface. In other words, it wraps around
> // batchedPlanNode's Next() method which advances full batches to
> ```

Does batchedPlanNode's Next() method do this? It looks to me like it panics.

---

*[pkg/sql/plan_batch.go, line 156 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L74fQMYE9IVZ05R8eXa:-L74fQMYE9IVZ05R8eXb:b2q6zgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/plan_batch.go#L156)):*
> ```Go
> //
> // This is an optimization upon serializeNode when it is known in
> // advance that the result rows will be discarded (e.g.  a
> ```

extra space between `e.g.` and `a`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-:-L74fjcWxN5XBxzZ6xCn:b1lh3ce)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 25 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/sql/delete.go, line 82 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L74XpFTGpWvjnXOTHKU:-L74hc9A60i9ZPAJwqGM:b645en5) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L82)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: these `/*comment*/` comments are diverging from our [style guide](https://github.com/cockroachdb/cockroach/blob/master/docs/style.md#inline-comments-for-arguments-to-function-calls), which expect spaces on either side of the comment. Are you ok with changing these (in a separate PR if you'd like) or should we explore amending the style guide?
</blockquote></details>

Yes a separate PR will be best. We have plenty of those throughout the code base.

---

*[pkg/sql/delete.go, line 116 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L74YqcnXh3-BN6ioLYo:-L74higE2Y_hOEzI4J7n:bhike5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L116)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is there a corresponding issue for this? Seems like it could be a good starter project.
</blockquote></details>

No idea - :microscope: :dog: here

---

*[pkg/sql/delete.go, line 134 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L74Z7sfZbnAgakVN-7A:-L74hwCJ3alryWKFFmiq:b2emgp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/delete.go#L134)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/to do rows/to do to rows/`
</blockquote></details>

Corrected, thanks.

---

*[pkg/sql/plan_batch.go, line 71 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L74cqWXlUIHjqTqPD9B:-L74iDw1CQud_AblyGPe:b-g3op08) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/plan_batch.go#L71)):*
> Note: Nodes that perform writes (e.g. INSERT) will not return from BatchedNext() before checking foreign key, uniqueness, and other CHECK constraints.

Works for me. Andrei wrote this, happy to take yours instead.

---

*[pkg/sql/plan_batch.go, line 93 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L74ey46GTP1Fo2wrqkR:-L74i0iL68xB0ZJnDiiF:b-j67258) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/plan_batch.go#L93)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does batchedPlanNode's Next() method do this? It looks to me like it panics.
</blockquote></details>

Typo in the comment; corrected. Thanks for spotting this.

---

*[pkg/sql/plan_batch.go, line 156 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-L74fQMYE9IVZ05R8eXa:-L74iQtO2oT77O7z2-4e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68575bd6e482df90bab48f60f274711b79e3f71c/pkg/sql/plan_batch.go#L156)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

extra space between `e.g.` and `a`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: avoid re-allocating a new rowIndexedVarContainer on every row inserted

Release note: None

knz commented:
Ok all statements are now properly batched so all the streaming-related problems are effectively fixed.

Now the only remaining problem is #20732. Working on a patch for that.

Then a commit to clean up the code a little (and remove dead code).

andreimatei commented:
LGTM

This all looks great to me! It'd be great if you found someone else to take a look too; I have to admit that some of the code in insert/update is over my head. It'd be great if you'd track down some of the authors.

---

Review status: 0 of 29 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/sql/tablewriter.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L70oEZZ6Fq0Gf1xHWCT:-L76x2rJxyeAz9mEHaSX:b29l2i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3193f03df098a4c087b6a9fa46f070ee83d73b/pkg/sql/tablewriter.go#L737)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It flushes the previous one and then allocates a new one. Isn't that the definition of rotation? We also "rotate" log files in the same way in the log package.
</blockquote></details>

"flushing" a batch is not a term we use. I see no analogy to log files :).

---

*[pkg/sql/tablewriter.go, line 187 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L77-M5fz8LpIw70ROZr:-L77-M5fz8LpIw70ROZs:b-hqnbxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91f0bf6f49fd113e1b97a527bc0dfeb1ea44ed5/pkg/sql/tablewriter.go#L187)):*
> ```Go
> 
> func (tu *tableUpdater) row(context.Context, tree.Datums, bool) (tree.Datums, error) {
> 	panic("unimplemented")
> ```

what's going on here with `row2`?

---

*[pkg/sql/update.go, line 495 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L76zkWWO7Xv0L8X4Ynp:-L76zkWWO7Xv0L8X4Ynq:b-apxyek) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91f0bf6f49fd113e1b97a527bc0dfeb1ea44ed5/pkg/sql/update.go#L495)):*
> ```Go
> 	// Now consume/accumulate the rows for this batch.
> 	lastBatch := false
> 	for {
> ```

this method is mighty long and also this code is super dense. I think splitting of a method for what comes after `entireRow := ...` would be great.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-:-L770MWb2VzqkUQY6E3o:b-ahrbdg)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 30 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/sql/tablewriter.go, line 187 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L77-M5fz8LpIw70ROZr:-L79m8aNRKhme42TAgDp:b-osh78u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91f0bf6f49fd113e1b97a527bc0dfeb1ea44ed5/pkg/sql/tablewriter.go#L187)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what's going on here with `row2`?
</blockquote></details>

The tablewriter interface is dead. I'll add a commit to clean this up.

---

*[pkg/sql/update.go, line 495 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L76zkWWO7Xv0L8X4Ynp:-L79mDUYHpA0PB0kURNG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91f0bf6f49fd113e1b97a527bc0dfeb1ea44ed5/pkg/sql/update.go#L495)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this method is mighty long and also this code is super dense. I think splitting of a method for what comes after `entireRow := ...` would be great.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373)*
<!-- Sent from Reviewable.io -->

knz commented:
I think the principal people who know the most about this now are @nvanbenschoten @justinj and @jordanlewis.
Feel free to chime in to check what is going on here.

---

Review status: 0 of 30 files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-:-L79mfk-ezDJPlvBqMwf:b26vet4)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: refactor the code for UPSERT; fix bugs

This patch applies the new pattern introduced by DELETE in a previous
patch and introduces proper batching behavior for UPSERT.

Perf difference:
```
name                                old time/op    new time/op    delta
SQL/Cockroach/Upsert/count=1-16        830µs ±16%     805µs ± 3%   ~     (p=1.000 n=5+5)
SQL/Cockroach/Upsert/count=10-16      1.11ms ±12%    1.10ms ± 7%   ~     (p=1.000 n=5+5)
SQL/Cockroach/Upsert/count=100-16     4.03ms ±11%    3.95ms ±14%   ~     (p=0.690 n=5+5)
SQL/Cockroach/Upsert/count=1000-16    27.8ms ±15%    25.4ms ± 2%   ~     (p=0.286 n=5+4)

name                                old alloc/op   new alloc/op   delta
SQL/Cockroach/Upsert/count=1-16        112kB ± 1%     112kB ± 1%   ~     (p=0.222 n=5+5)
SQL/Cockroach/Upsert/count=10-16       154kB ± 1%     154kB ± 1%   ~     (p=0.690 n=5+5)
SQL/Cockroach/Upsert/count=100-16      710kB ± 1%     708kB ± 1%   ~     (p=1.000 n=5+5)
SQL/Cockroach/Upsert/count=1000-16    6.22MB ± 2%    6.23MB ± 2%   ~     (p=1.000 n=5+5)

name                                old allocs/op  new allocs/op  delta
SQL/Cockroach/Upsert/count=1-16          988 ± 1%       994 ± 1%   ~     (p=0.175 n=5+5)
SQL/Cockroach/Upsert/count=10-16       1.34k ± 2%     1.35k ± 2%   ~     (p=0.317 n=5+5)
SQL/Cockroach/Upsert/count=100-16      5.27k ±16%     5.09k ± 9%   ~     (p=0.841 n=5+5)
SQL/Cockroach/Upsert/count=1000-16     44.4k ±14%     43.8k ±15%   ~     (p=1.000 n=5+5)
```

(negligible)

This also enables deleting the `returningHelper` and the `editNode`
infrastructure, which is not needed any more.

Release note (bug fix): Removed a limitation where `UPSERT` would fail
if the number of modified rows was too large.

Release note (bug fix): `UPSERT` now properly reports the count of
modified rows when `RETURNING` is not specified.

knz commented:
Review status: 0 of 29 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/sql/tablewriter.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373#-L70oEZZ6Fq0Gf1xHWCT:-L7EcFD4Zy2hAD0KOAk9:bi5ypcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d3193f03df098a4c087b6a9fa46f070ee83d73b/pkg/sql/tablewriter.go#L737)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

"flushing" a batch is not a term we use. I see no analogy to log files :).
</blockquote></details>

OK, changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23373)*
<!-- Sent from Reviewable.io -->

knz commented:
I rebased on top of the spooling patch however doing the work in that order created some mess which I now need to resolve: we're introducing a spool on the mutation statement itself, but if there's a RETURNING clause this is now handled with an additional `render` stage. That is currently placed outside of the spool, which while technically correct is highly inefficient for top-level mutations.

knz commented:
The solution is to adapt the logic in `expand_plan` that adds the spool to have it add it at the right place. Will do that a bit later. Since we're not merging this for 2.0 I guess there's not a rush.

andreimatei commented:
On the [known limitations](https://www.cockroachlabs.com/docs/dev/known-limitations.html) page there's a reference to the number of writes a txn can make (which I'm currently updating to talk about statements, not txns). When this PR goes in, the limitation should be removed or relaxed.

---

Review status: 0 of 37 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-:-L97lQHm3rT-Py8crQPl:b-wg98fa)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Como esta progresando el PR, mi amigo? 2.0.1 is coming.

---

Review status: 0 of 37 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23373#-:-L9fmfmo5RHdYKRslOWa:bjxwldj)*
<!-- Sent from Reviewable.io -->

knz commented:
Rebased and adapted post-2.0.
Still need to resolve the spooling issue. Right now a plan looks like this, and it's ugly:

```
root@:26257/t> explain select * from [delete from kv returning k+v] where @1 = 3;
+-------------------------------+-------+-------------+
|             Tree              | Field | Description |
+-------------------------------+-------+-------------+
| filter                        |       |             |
|  └── render                   |       |             |
|       └── spool               |       |             |
|            └── run            |       |             |
|                 └── delete    |       |             |
|                      │        | from  | kv          |
|                      └── scan |       |             |
|                               | table | kv@primary  |
|                               | spans | ALL         |
+-------------------------------+-------+-------------+
(9 rows)
```

That is, the renders of a RETURNING clause are performed outside of the spool, not inside. This is not incorrect but it is wasteful.

knz committed:
sql: move tableDeleter to its own file

Release note: None

knz committed:
sql: move tableUpdater to its own file

Release note: None

knz committed:
sql: move tableInserter to its own file

Release note: None

knz committed:
sql: move tableUpserter and fastTableUpserter to their own files

Release note: None

knz committed:
sql: extend the logic test mechanism to test rows affected counts

This patch introduces the logic test syntax `statement count N` that
is like `statement ok` but also checks that the final RowsAffected
count is equal to N.

It also introduces the relevant DELETE/UPSERT/INSERT/UPDATE tests,
which exercise this code path.

Release note: None

knz committed:
sql: optimize the placement of spools

Prior to this patch, the spool stage if needed would be added just
before the mutation stage that requires it. Then a separate condition
would remove the spool stage if it was at the top level of a logical
plan

With the new handling of RETURNING however, this would cause the
unfortunate situation where a top level mutation _statement_ with
RETURNING would get a spool although it is not needed (reminder:
top-level mutations do not need a spool; a spool is only required if
there is further processing by a surrounding query) -- this is because
RETURNING has been recently modified to be handled by a regular render
(projection) stage after the mutation stage, instead of being executed
inside the mutation stage itself. So a plan would look like returning
- spool - mutation, whereas we'd really like to see returning -
mutation directly.

This patch solves the problem by introducing two plan transforms that
"pull up" a spool:

- render - spool is replaced by spool - render
- filter - spool is replaced by spool - filter
- distinct - spool is replaced by spool - distinct

The first one addresses the immediate non-optimal situation with
RETURNING that required this patch. The other two are generally-safe,
generally-desirable transform that reduce the amount or rows stored
inside the spool.

Release note (sql change): a minor optimization is introduced that
reduces the amount of RAM used when a query performs further
computations on the result of a mutation statement
(INSERT/DELETE/UPSERT/UPSERT) combined with RETURNING.

knz commented:
There we go, fixed it.

knz committed:
sql: add a test for the validity of values produced by RETURNING

As detected in issue #22304, a mutation with RETURNING could let
values go through to the client that would violate constraints. This
was fixed in a combination of previous patches; this commit adds the
corresponding regression test.

Release note: None

knz commented:
Also now with a regression test for the original issue (#22304) that motivated all this work.

knz commented:
bors r+

andreimatei commented:
@knz I'm ecstatic to see this merged! Would you mind updating this doc? https://www.cockroachlabs.com/docs/dev/known-limitations.html#write-and-update-limits-for-a-single-statement

knz commented:
done, thanks!
